service/matching/taskReader.go

```diff
@@ -123,22 +125,21 @@ dispatchLoop:
 			}
 			task := newInternalTask(taskInfo, tr.tlMgr.completeTask, enumsspb.TASK_SOURCE_DB_BACKLOG, "", false)
 			for ctx.Err() == nil {
-				// We checked if the task was expired before putting it in the buffer, but it
-				// might have expired while it sat in the buffer, so we should check again.
-				if taskqueue.IsTaskExpired(taskInfo) {
+				if valid := tr.taskValidator.maybeValidate(taskInfo, tr.tlMgr.taskQueueID.taskType); !valid {
 					task.finish(nil)
 					tr.taggedMetricsHandler().Counter(metrics.ExpiredTasksPerTaskQueueCounter.GetMetricName()).Record(1)
 					// Don't try to set read level here because it may have been advanced already.
-					break
+					continue dispatchLoop
 				}
-				err := tr.tlMgr.engine.DispatchSpooledTask(ctx, task, tr.tlMgr.taskQueueID, tr.tlMgr.stickyInfo)
+
+				taskCtx, cancel := context.WithTimeout(ctx, taskReaderOfferTimeout)
+				err := tr.tlMgr.engine.DispatchSpooledTask(taskCtx, task, tr.tlMgr.taskQueueID, tr.tlMgr.stickyInfo)
+				cancel()
 				if err == nil {
-					break
-				}
-				if err == context.Canceled {
-					break dispatchLoop
+					continue dispatchLoop
 				}
-				// this should never happen unless there is a bug - don't drop the task
+
+				// if task is still valid (truly valid or unable to verify if task is valid)
```

Comment:
This comment is extremely confusing. Can you elaborate on what you mean here? It looks like this is actually hit when we failed DispatchSpooledTask, so I'm not sure what that has to do with the validity of the task

Reply:
about the logic:
if we are 100% sure the task is invalid, then it is safe to not dispatch the task (by discarding the task)
if we are 100% sure the task is valid (can be started), then the logic must dispatch the task, until the task is 100% invalid
if we are not sure whether the task is valid or invalid, then the logic must dispatch the task, until the task is 100% invalid


about the use case (activity task):
1. namespace is active in cluster A, and also replicates to cluster B
2. for activity schedule event, there is a corresponding activity transfer task generated in history service
3. in cluster A: since cluster A is active, cluster A's history service can immediately execute the activity transfer task by push it to matching service
  NOTE: there is no guarantee that all matching tasks will be poll by SDK immediately, e.g. an activity task can be marked as started few minutes after scheduled, or never
4. in cluster B: since cluster B is passive, cluster B's history service can only in mem *verify* the activity transfer task by waiting for the activity started event.
5. since there is no guarantee an activity start event will ever happen (see # 3.NOTE), and we cannot let this activity transfer task stuck the history service (see # 4), cluster B's history service will eventually push the activity transfer task to matching (so task is not lost)

above behavior will slowly accumulate lots of invalid tasks on passive side matching service. this backlog of invalid tasks need to be cleaned up properly.